include/ast.hpp

```diff
@@ -220,17 +220,34 @@ struct CompoundStmtNode : public StmtNode {
   std::vector<std::unique_ptr<StmtNode>> stmts;
 };
 
-/// @brief Root of the entire program.
-struct ProgramNode : public AstNode {
+/// @brief An external declaration can be a definition of a function or an
+/// object.
+struct ExternDeclNode : public AstNode {
+  ExternDeclNode(
+      Location loc,
+      std::variant<std::unique_ptr<FuncDefNode>, std::unique_ptr<DeclStmtNode>>
+          extern_decl)
+      : AstNode{loc}, extern_decl{std::move(extern_decl)} {}
+
+  void Accept(NonModifyingVisitor&) const override;
+  void Accept(ModifyingVisitor&) override;
+
+  std::variant<std::unique_ptr<FuncDefNode>, std::unique_ptr<DeclStmtNode>>
+      extern_decl;
```

Comment:
Maybe just `decl` will suffice? Since the struct is already named `ExternDeclNode`, accessing the field as `extern_decl.extern_decl` seems redundant.

```suggestion
  std::variant<std::unique_ptr<FuncDefNode>, std::unique_ptr<DeclStmtNode>> 
      decl;
```